File: haskell/day1/app/Main.hs
module Main where

import Data.Text (pack, Text, splitOn, null) 
import Data.Text.Read (decimal, Reader)
import Data.List (sortBy)

main :: IO ()
main = do 
  input <- readFile "../../input/day1.txt"
  mapM_ (day1Solution input) [1,3]

day1Solution :: String -> Int -> IO () 
day1Solution input n = printResult n . fmap (getNTopCalories n) . traverse (fmap sumCalories . parseCalories) . getListPerElf . pack $ input

printResult :: Int -> Either String Integer -> IO () 
printResult n result = do 
  case result of 
    Left err -> putStrLn $  "Error when calculating top " ++ show n ++ " elves calories: " ++ err
    Right max_cals -> putStrLn $ "Calories of top " ++ show n ++ " elves calories: " ++ show max_cals

nonEmptySplitOn :: Text -> Text -> [Text]
nonEmptySplitOn on = filter (not . Data.Text.null) . splitOn on

getListPerElf :: Text -> [Text]
getListPerElf = nonEmptySplitOn (pack "\n\n")

parseCalories :: Text -> Either String [Integer]
parseCalories = traverse (fmap fst . decimal) . nonEmptySplitOn (pack "\n")

sumCalories :: [Integer] -> Integer 
sumCalories = sum

getNTopCalories :: Int -> [Integer] -> Integer
getNTopCalories n = sum . take n . sortBy (flip compare)

File: rust/day1/src/main.rs
use anyhow::{Context, Result};
use std::path::Path;

fn main() -> Result<()> {
    std::fs::read_to_string(Path::new("../../input/day1.txt"))
        .context("Couldn't read input!")
        .map(|input| {
            [1, 3]
                .into_iter()
                .map(|top_n| {
                    parsed_summed_up_calories(&input)
                        .map(sort_calories)
                        .map(get_top_n_calories(top_n))
                        .map(|calories| (calories, top_n))
                })
                .for_each(print_result)
        })
}

#[derive(Default, Debug)]
struct Calories(pub usize);

// utils

fn str_not_empty(str: &&str) -> bool {
    !str.is_empty()
}

// printing

fn print_result(result: Result<(Calories, usize)>) {
    match result {
        Ok((calories, n)) => println!("Top {n} elves carry {calories:?}"),
        Err(err) => println!("Algo failed with error: {err:?}"),
    }
}

// parsing & preparation

fn parsed_summed_up_calories(input: &str) -> Result<Vec<Calories>> {
    input
        .split("\n\n")
        .filter(str_not_empty)
        .map(sum_calories)
        .collect()
}

fn sum_calories(list_per_elf: &str) -> Result<Calories> {
    list_per_elf
        .split('\n')
        .filter(str_not_empty)
        .map(parse_calories)
        .sum::<Result<usize>>()
        .map(Calories)
}

fn parse_calories(calories_str: &str) -> Result<usize> {
    calories_str
        .parse::<usize>()
        .context("Didn't find parseable calories!")
}

// sorting

fn sort_calories(mut elves_calories: Vec<Calories>) -> Vec<Calories> {
    elves_calories.sort_by(|Calories(ref a), Calories(ref b)| a.cmp(b).reverse());
    elves_calories
}

// calculating top sum

fn get_top_n_calories(top_n: usize) -> impl FnOnce(Vec<Calories>) -> Calories {
    move |elves_calories| {
        elves_calories
            .iter()
            .take(top_n)
            .fold(Calories::default(), |Calories(sum), Calories(cals)| {
                Calories(sum + cals)
            })
    }
}
